pelican/plugins/_utils.py

```diff
@@ -107,3 +108,23 @@ def load_plugins(settings):
         plugins = list(namespace_plugins.values())
 
     return plugins
+
+
+def get_plugin_name(plugin):
+    """
+    Plugins can be passed as module objects, however this breaks caching as
+    module objects cannot be pickled. To work around this, all plugins are
+    stringified post-initialization.
+    """
+    if isinstance(plugin, str):
```

Comment:
These special cases are required to keep the function as sketched in your comment from returning `'str'` or `'type'`.

Reply:
Well, for one, this should not receive a `str` at all. It will be passed loaded plugins which will be modules or similar plugin-like objects. And current conditions are a bit wrong. imported modules would not return correct values for example. Class as an input is a bit unusual but I guess you can force a class to behave like plugin (if you throw around `@staticmethod`s). Normal classes (like the one you wrote for test) would require an instance of it. Still we can account for a class.

So the values would be:
* `plugin.__name__` if it's a module
* `plugin.__qualname__` if it's a class (`if inspect.isclass(plugin):` would correspond to this, so it's not exactly correct right now)
* `plugin.__class__.__qualname__` (or `type(plugin).__qualname__`) if it's an instance. I think this will always succeed, so this should also cover everything else

Which could look like:
```python
def get_plugin_name(plugin):
    if inspect.ismodule(plugin):
        return plugin.__name__

    if inspect.isclass(plugin):
        return plugin.__qualname__

    return type(plugin).__qualname__
```